### PR TITLE
Abstract QL connection code and create new MySQLi connection

### PR DESCRIPTION
**Pull request by TomK**

### CODE REVIEW

File: src/Ql/AbstractQlConnection.php [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+   *
+   * @return bool
+   */
+  private function _shouldReconnectAfterException(\Exception $e)
+  {
+    // 2006  = MySQL server has gone away
+    // 1047  = ER_UNKNOWN_COM_ERROR - happens when a PXC node is resyncing:
+    //          "WSREP has not yet prepared node for application use"
+    // HY000 = General SQL error
+    $codes = ['2006', '1047', '2014', 'HY000'];
```
Comment (rgooding) [inline src/Ql/AbstractQlConnection.php]: What is code 2014? Can you add it to the comment so we don't forget?

File: src/Ql/MySQLiConnection.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
+<?php
+namespace Packaged\Dal\Ql;
+
+use Packaged\Dal\Exceptions\Connection\ConnectionException;
+use Packaged\Helpers\ValueAs;
+
+/**
+ * Class PdoConnection
```
Comment (rgooding) [inline src/Ql/MySQLiConnection.php]: It's MySQLiConnection, not PdoConnection

File: src/Ql/MySQLiConnection.php [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+  protected function _freeStatement($stmt)
+  {
+    if($stmt instanceof \mysqli_stmt)
+    {
+      $stmt->free_result();
+    }
+    else
+    {
+      throw new ConnectionException('Incorrect type passed to close');
```
Comment (rgooding) [inline src/Ql/MySQLiConnection.php]: "Incorrect type passed to _freeStatement"?

File: src/Ql/PdoConnection.php [hunk 4]
```diff
@@ -1,130 +1,69 @@
...
-        try
-        {
-          $options = array_replace(
-            $this->_defaultOptions(),
-            ValueAs::arr($this->_config()->getItem('options'))
-          );
+      $dsn = sprintf(
+        "mysql:host=%s;dbname=%s;port=%d",
+        $this->_config()->getItem('hostname', '127.0.0.1'),
+        $this->_selectedDb ?: $this->_config()->getItem('database', ''),
```
Comment (rgooding) [inline src/Ql/PdoConnection.php]: Does this work if the database name is empty? (i.e. can you put `mysql:host=127.0.0.1;dbname=;port=3306` in the DSN?)

File: src/Ql/AbstractQlConnection.php (hunk 1 above)
Comment (TomK) [reply src/Ql/AbstractQlConnection.php]: good catch, i think that one is a mistake. It's "bad query", we don't need to reconnect for that.

**Pull request merged**